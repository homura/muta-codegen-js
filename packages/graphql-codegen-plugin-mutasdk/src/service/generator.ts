import { upperFirst } from 'lodash';
import { ServiceDef } from '../types';

function formatTypeName(type: string, defaults = 'null') {
  return type ?? defaults;
}

export function generate(serviceDef: ServiceDef): string {
  const { name, query, mutation } = serviceDef;
  const serviceClassName = upperFirst(name);
  const template = `
export const ${serviceClassName} = createBindingClass(
  '${name}',
  {
    ${query
    .map(
      ({ name, payloadType, returnType }) =>
        // prettier-ignore
        `${name}: query<${formatTypeName(payloadType)}, ${formatTypeName(returnType)}>(),`,
    )
    .join('\n    ')}
  },
  {
    ${mutation
    .map(
      ({ name, payloadType, returnType }) =>
        // prettier-ignore
        `${name}: mutation<${formatTypeName(payloadType)}, ${formatTypeName(returnType)}>(),`,
    )
    .join('\n    ')}
  }
);
`;
  return template;
}
