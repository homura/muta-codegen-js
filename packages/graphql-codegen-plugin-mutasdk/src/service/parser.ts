import {
  GraphQLField,
  GraphQLInputObjectType,
  GraphQLNonNull,
  GraphQLScalarType,
  GraphQLSchema,
  isNonNullType,
} from 'graphql';
import { MethodDef } from '../types';

function fieldToMethodDef(field: GraphQLField<any, any>): MethodDef {
  const payloadType = field.args[0].type;
  const payload = isNonNullType(payloadType)
    ? payloadType.ofType
    : (payloadType as GraphQLScalarType | GraphQLInputObjectType);

  return {
    name: field.name,
    payloadType: payload?.name,
    returnType: (field.type as GraphQLNonNull<any>)?.ofType?.name,
  };
}

export function parseMethodDef(
  schema: GraphQLSchema,
): { query: MethodDef[]; mutation: MethodDef[] } {
  const queryFields = schema.getQueryType()?.getFields() ?? [];
  const query = Object.values(queryFields).map(fieldToMethodDef);

  const mutationType = schema.getMutationType()?.getFields() ?? [];
  const mutation = Object.values(mutationType).map(fieldToMethodDef);

  return {
    query,
    mutation,
  };
}
