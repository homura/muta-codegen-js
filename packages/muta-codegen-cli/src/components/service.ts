import * as typescriptPlugin from '@graphql-codegen/typescript';
import { service as servicePlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateServiceBindingCode(
  serviceName: string,
  schemaSource: string,
): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      service: servicePlugin,
    },
    plugins: [{ service: { service: serviceName } }, { typescript: {} }],
    schemaSource,
  });
}
