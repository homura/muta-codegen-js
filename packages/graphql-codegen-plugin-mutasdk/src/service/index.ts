import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import { generate } from './generator';
import { parseMethodDef } from './parser';

const importSegment = `
import { createBindingClass, read, Read, write, Write } from "@mutajs/service";
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec } from "@mutajs/types";
`;

type ServiceCodegenConfig = { service: string };

export const plugin: PluginFunction<ServiceCodegenConfig> = (
  schema,
  documents,
  config,
) => {
  const { query, mutation } = parseMethodDef(schema);

  return {
    prepend: [importSegment],
    content: generate({
      mutation,
      name: config.service,
      query,
    }),
  };
};
