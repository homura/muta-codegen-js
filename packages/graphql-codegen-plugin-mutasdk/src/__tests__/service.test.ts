import { buildSchema } from 'graphql';
import { generate } from '../service/generator';
import { parseMethodDef } from '../service/parser';

const source = `
  schema {
  query: Query
  mutation: Mutation
}

type Query {
  getAsset(payload: GetAssetPayload!): Asset!
  getBalance(payload: GetBalancePayload!): u64!
}

input GetAssetPayload {
  id: Hash!
}

input GetBalancePayload {
  id: Hash!
  user: Address!
}

# The field in mutation strictly corresponds to a transaction, therefore,
# multiple mutations executed at the same time should not be allowed
type Mutation {
  transfer(payload: TransferPayload!): Null!
  createAsset(payload: CreateAssetPayload!): Asset!
}

scalar Null

input TransferPayload {
  id: Hash!
  user: Address!
}

input CreateAssetPayload {
  supply: u64!
  name: String!
  symbol: String!
}

type GeneralMutationResponse {
  txHash: Hash!

  """
  Since receipt is executed asynchronously,
  it is impossible to keep the connection for a long time,
  so only the \`TYPE\` will be returned during mutation.
  The \`receiptType\` would always be null in response
  """
  receiptType: String
}
type CreateAssetReceipt {
  supply: u64!
  name: String!
  symbol: String!
  id: Hash!
}
type CreateAssetMutationResponse {
  txHash: Hash!
  receiptType: CreateAssetReceipt
}
type Asset {
  id: Hash!
  name: String!
  symbol: String!
  supply: u64!
  issuer: Address!
}
scalar Address
scalar u64
scalar Hash
`;

it('parse asset service', () => {
  expect(parseMethodDef(buildSchema(source))).toEqual({
    query: [
      //getAsset(payload: GetAssetPayload): Asset!
      {
        name: 'getAsset',
        payloadType: 'GetAssetPayload',
        returnType: 'Asset',
      },
      // getBalance(payload: GetBalancePayload): u64!
      {
        name: 'getBalance',
        payloadType: 'GetBalancePayload',
        returnType: 'u64',
      },
    ],
    mutation: [
      //  transfer(payload: TransferPayload): Null
      {
        name: 'transfer',
        payloadType: 'TransferPayload',
        returnType: 'Null',
      },
      // createAsset(payload: CreateAssetPayload ): Asset!
      {
        name: 'createAsset',
        payloadType: 'CreateAssetPayload',
        returnType: 'Asset',
      },
    ],
  });
});

it('generate empty service', () => {
  expect(
    generate({
      name: 'asset',
      query: [],
      mutation: [],
    }),
  ).toMatchSnapshot();
});

it('generate AssetService binding', () => {
  expect(
    generate({
      name: 'asset',
      query: [
        {
          name: 'get_asset',
          payloadType: 'GetAssetPayload',
          returnType: 'Asset',
        },
        {
          name: 'get_balance',
          payloadType: 'GetBalancePayload',
          returnType: 'Balance',
        },
      ],
      mutation: [
        {
          name: 'transfer',
          payloadType: 'TransferPayload',
          returnType: 'Null',
        },
      ],
    }),
  ).toMatchSnapshot();
});
