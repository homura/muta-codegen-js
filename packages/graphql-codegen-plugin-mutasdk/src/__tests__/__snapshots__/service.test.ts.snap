// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate AssetService binding 1`] = `
"
export const Asset = createBindingClass(
  'asset',
  {
    get_asset: query<GetAssetPayload, Asset>(),
    get_balance: query<GetBalancePayload, Balance>(),
  },
  {
    transfer: mutation<TransferPayload, Null>(),
  }
);
"
`;

exports[`generate empty service 1`] = `
"
export const Asset = createBindingClass(
  'asset',
  {
    
  },
  {
    
  }
);
"
`;
